lib/code_corps/github.ex

```diff
@@ -1,41 +1,245 @@
+  end
+
+  defmodule APIError do
+    defstruct [:documentation_url, :errors, :message, :status_code]
+
+    @type t :: %__MODULE__{
+      documentation_url: String.t | nil,
+      errors: List.t | nil,
+      message: String.t | nil,
+      status_code: pos_integer | nil
+    }
+
+    @spec new({integer, map}) :: t
+    def new({status_code, %{"message" => message, "errors" => errors}}) do
+      errors = Enum.into(errors, [], fn error -> convert_error(error) end)
+
+      %__MODULE__{
+        errors: errors,
+        message: message,
+        status_code: status_code
+      }
+    end
+
+    @spec new({integer, map}) :: t
+    def new({status_code, %{"message" => message, "documentation_url" => documentation_url}}) do
+      %__MODULE__{
+        documentation_url: documentation_url,
+        message: message,
+        status_code: status_code
+      }
+    end
+
+    @spec new({integer, map}) :: t
+    def new({status_code, %{"message" => message}}) do
+      %__MODULE__{
+        message: message,
+        status_code: status_code
+      }
+    end
+
+    defp convert_error(%{"code" => code, "field" => field, "resource" => resource}) do
+      APIErrorObject.new([code: code, field: field, resource: resource])
+    end
+  end
+
+  defmodule HTTPClientFailed do
+    defstruct [:reason, message: """
+    The GitHub HTTP client encountered an error while communicating with
+    the GitHub API.
+    """]
+
+    def new(opts) do
+      struct(__MODULE__, opts)
+    end
+  end
+
+  @type method :: :get | :post | :put | :delete | :patch
+  @type headers :: %{String.t => String.t} | %{}
+  @type body :: {:multipart, list} | map
+  @typep http_success :: {:ok, integer, [{String.t, String.t}], String.t}
+  @typep http_failure :: {:error, term}
+
+  @type api_error_struct ::
+    %APIError{}            |
+    %HTTPClientFailed{}
+
+  @spec get_base_url() :: String.t
+  defp get_base_url() do
+    Application.get_env(:code_corps, :github_base_url) || "https://api.github.com/"
+  end
+
+  #
+  @spec get_token_url() :: String.t
+  defp get_token_url() do
+    Application.get_env(:code_corps, :github_oauth_url) || "https://github.com/login/oauth/access_token"
+  end
+
+  @spec use_pool?() :: boolean
+  defp use_pool?() do
+    Application.get_env(:code_corps, :github_api_use_connection_pool)
+  end
+
+  @spec add_default_headers(headers) :: headers
+  defp add_default_headers(existing_headers) do
+    Map.merge(%{"Accept" => "application/vnd.github.machine-man-preview+json"}, existing_headers)
+  end
+
+  @spec add_access_token_header(headers, String.t | nil) :: headers
+  defp add_access_token_header(existing_headers, nil), do: existing_headers
+  defp add_access_token_header(existing_headers, access_token) do
+    Map.put(existing_headers, "Authorization", "token #{access_token}")
+  end
+
+  defp generate_jwt do
+    # generates JWT with RSA key, using 10 minute expires time, RS256 algo,
+    # and issuer of the GitHub app's id
+    "$JWT"
+  end
```

Comment:
This needs to be done still. Not sure best way to handle the RSA key portion, either.